### PR TITLE
Fix ttnn.sort wrong indices for bf16

### LINKED ISSUE #46331: [TTIRToTTNN] Integer ttir.SortOp routed through ttnn.sort compares keys in bf16, returning wrong sort indices
## Describe the bug

`ttnn.sort` returns a **correct sorted-values tensor but an incorrect `indices` tensor** whenever the sort keys are integers whose magnitude exceeds ~256. The bitonic sort reuses the TopK SFPU sort LLK, which compares keys in **bf16** (8-bit mantissa). Keys that are not exactly representable in bf16 (e.g. distinct integers `> 256`) collide into ties, and because the `stable` flag is ignored, the indices of the tied keys are ordered arbitrarily.

The result: `ttnn.sort(...).values` looks correct (swapping two equal-in-bf16 values keeps the row "sorted"), but `ttnn.sort(...).indices`i.e. the `argsort` permutation is wrong. The dtype-cast operand workaround (`createSortOpOperandsWorkarounds`, casts integer input → `ui16`) only makes the op *run*; it does not fix the comparison precision, so the indices are still wrong.

## Downstream impact

Breaks `qwen_2_5_vl/pytorch-3B_Instruct` inference in tt-xla. The vision tower undoes its window-attention reordering with `reverse = torch.argsort(window_index); merged = merged[reverse]`, where `window_index` is an `int64` permutation of `0..550`. The wrong `argsort` indices scatter every image-patch embedding to the wrong token position, giving full-model logits **PCC ≈ 0.24** (required ≥ 0.99). Any model that uses sort/argsort **indices** over integer or `>256`-valued keys is affected.

## Root cause

`sort_single_row_single_core.cpp` does the compare/swap with the TopK SFPU LLK (`ckernel::topk_local_sort` / `topk_merge`), which compares keys in **bf16**, so distinct integers `> 256` collide into ties and since the indices ride the same compare/swap network, those ties scramble the indices.

## Steps to reproduce

In tt-xla (the integer-argsort regression test):

```
git checkout sonalib/argsort
pytest -svv tests/torch/ops/test_sort.py
```

## Sanity Test

Minimal `ttnn`-level reproducer (no tt-xla / MLIR needed):

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

def check(n, hi, torch_dtype):
    # distinct keys: a permutation of [0, n) spread into [0, hi)
    perm = torch.randperm(n)
    x = (perm.to(torch.float64) / max(n - 1, 1) * hi).to(torch_dtype).reshape(1, n)

    tt = ttnn.from_torch(x, device=device, layout=ttnn.TILE_LAYOUT)
    _, tt_idx = ttnn.sort(tt, dim=-1)
    dev_idx = ttnn.to_torch(tt_idx).to(torch.int64)

    golden = torch.argsort(x.to(torch.float64), dim=-1)
    match = (dev_idx == golden).float().mean().item()
    print(f"n={n:4d} hi={hi:4.0f} {str(torch_dtype):16s} index_match={match:.4f}")

check(256, 255, torch.float32)   # values < 256  -> expect ~1.0  
check(512, 511, torch.float32)   # values > 256 
check(551, 550, torch.int32)     # qwen window_index size    

ttnn.close_device(device)
```

### PR DESCRIPTION
**Pull request by abdulla-TT**
### PR Category

Bug fix

### Summary

Fixes #46331 — `ttnn.sort` returns wrong indices for integer keys larger than ~256.

**Root cause.** The sort's TopK SFPU LLK (`ckernel::topk_local_sort`) compares tile values in **bf16** by default. bf16 has only an 8-bit mantissa, so integers above 256 (e.g. `257`, `258`, `259`) all round to the same bf16 value and become equal to the comparator. The comparator then leaves the two tied elements in whatever order the merge-sort network happened to place them, so the returned indices are non-deterministic and can be wrong for any input with:
* dtype `UINT16` or `FLOAT32` outside the exact-bf16 range, or
* dtype `BFLOAT16` where the user's numeric range collides after truncation.

Prior workaround (`createSortOpOperandsWorkarounds`, casting integer input → `ui16` on host) only kept the op runnable; the comparison precision was still bf16, so the **indices** stayed wrong.

**Fix.** Sort keys are now compared as `Float32` on the SFPU:
* `FLOAT32` input: unchanged path — already ran the sort kernel with `fp32_dest_acc_en = true`.
* `UINT16` input: promoted to `Float32` for the sort:
  * The reader kernel DMAs raw `UInt16` tiles into a staging CB (`c_11` / `c_12`) and software-converts each element to `Float32` on the Baby RISC-V core into a `Float32` primary CB (`c_0`). All values in `[0, 65535]` are exactly representable in Float32's 24-bit mantissa, so the comparison is exact. Hardware unpack cannot numerically convert `UInt16 → Float32` per the Tensix ISA (only `UInt16 → UInt16` dest is supported), which is why the cast is done on the RISC-V core with an explicit `__sync_synchronize()` fence before `push_back` so the compute kernel sees the completed stores.
  * The writer kernel does the reverse: it software-converts the sorted `Float32` values back to `UInt16` in a staging CB, then DMAs the result to DRAM.
  * Because `fp32_dest_acc_en = true` forces the sort kernel to use `INT32` (4 KB) index tiles, `UINT16` inputs now return **`UINT32`** indices (matching what `FLOAT32` inputs already did).
* `BFLOAT16` input: no behavioral change (already the bf16-comparison path; users needing exact-int index guarantees should pass `UINT16` or `FLOAT32`).

**Additional coverage added along the way** while wiring the `Float32` conversion through all three program factories:
* `SortProgramFactorySingleRowSingleCore` — reader/writer conversion + fence for both TILE and ROW_MAJOR layouts.
* `SortProgramFactorySingleRowMultiCore` — reader/writer conversion for both TILE and ROW_MAJOR layouts, plus a fix for a latent DMA sizing bug in the coordinator's passthrough (used `get_tile_size(c_0)` which is now `Float32` = 4 KB when the DRAM tiles are `UInt16` = 2 KB).
* MultiCore ROW_MAJOR path was rewritten to use a pair-row CB layout (each `c_6` / `c_8` page holds one row of both tiles side-by-side) so `tilize_block(cb, 2)` produces the expected 2-tile output. This fixes a pre-existing layout bug in the MultiCore RM path.
* `SortDeviceOperation::select_program_factory` — `UINT16` inputs above the SingleCore threshold now route to `SortProgramFactorySingleRowMultiCore` (the CrossCore factory does not yet implement the `UInt16 ↔ Float32` conversion CBs).
* `UINT16` + ROW_MAJOR uses a lower SingleCore threshold (`Wt <= 32`), because promoting the SingleCore RM value CBs to `Float32` doubles per-row storage and `Wt = 64` would OOM ~2 MB in a 1.5 MB L1. Larger `Wt` routes to MultiCore, which uses per-tile pages independent of `Wt`.

### Notes for reviewers

* The comparison-precision fix is entirely on-device — no host-side typecast, no change to the public API surface other than the `UINT16 → UINT32` index dtype (which matches the pre-existing `FLOAT32` input contract).
* `UINT16` padding uses `UINT16_MAX` (ascending) / `0` (descending) as sentinels because `UInt16` cannot represent `±inf`. This has a **known limitation** when the input contains those exact values AND the last dim needs W-padding to the next power of two: since the sort is unstable, a padded element can tie with the real value at the slice boundary and yield an index outside the original last dim. Follow-up work is to have the reader kernel override the padded region with a `Float32` sentinel keyed off `original_last_dim`.
* Files most worth focusing on:
  * `sort_program_factory.cpp` — CB format promotion, `fp32_dest_acc_en`, `UnpackToDestFp32`, new staging CBs.
  * `reader_single_row_{single,multi}_core.cpp`, `writer_single_row_{single,multi}_core.cpp` — RISC-V `UInt16 ↔ Float32` conversion loops with `__sync_synchronize()` fences.
  * `sort_device_operation.cpp` — routing threshold changes for `UINT16` inputs.
  * `tests/ttnn/unit_tests/operations/data_movement/test_sort.py` — regression coverage: `test_sort_uint16_index_correctness`, `test_sort_uint16_row_major_correctness`, `test_sort_uint16_index_correctness_multicore` covering the ranges that used to alias in bf16.
* Full `test_sort.py` suite: **125 passed** locally on WH.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:abdulla/sort_wrong_indices_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:abdulla/sort_wrong_indices_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:abdulla/sort_wrong_indices_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:abdulla/sort_wrong_indices_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:abdulla/sort_wrong_indices_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:abdulla/sort_wrong_indices_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:abdulla/sort_wrong_indices_fix)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:abdulla/sort_wrong_indices_fix)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/data_movement/test_sort.py [hunk 1]
```diff
@@ -304,6 +305,152 @@ def test_sort_datatypes(shape, dim, descending, torch_value_dtype, ttnn_value_dt
...
+    "n, hi, descending",
+    [
+        # ROW_MAJOR UINT16 sort:
+        #   Wt = ceil(n / 32) after W-padding.  n <= 2048 (Wt <= 64) routes to
+        #   SingleCore; n > 2048 (Wt > 64) routes to MultiCore, which now also
+        #   does the UInt16↔Float32 conversion in its ROW_MAJOR reader/writer.
+        (64, 63, False),  # Wt = 2, SingleCore RM
+        (551, 550, False),  # Wt = 18, SingleCore RM
+        (551, 550, True),
+        # Wt = 64 SingleCore RM UINT16 hits L1 OOM (~2 MB CBs) — pre-existing
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/data_movement/test_sort.py]: With `SORT_WT_THRESHOLD_UINT16_ROW_MAJOR = 32`, Wt=64 UINT16 RM no longer reaches SingleCore — `select_program_factory` sends it to MultiCore — so the OOM justification no longer holds and this case should pass. Wt=64 (n in 1025..2048) is the first width the new threshold reroutes, i.e. the one boundary worth covering: consider adding `(2048, 2047, False)`.

Separately, the Wt numbers in these comments ignore the power-of-two W padding in `pre_sort_transform_tensor`: n=551 pads to 1024 (Wt=32, not 18) and n=2080 pads to 4096 (Wt=128, not 65). Wt is always a power of two, so Wt=65 cannot occur.

File: ttnn/cpp/ttnn/operations/data_movement/sort/device/sort_program_factory.cpp [hunk 2]
```diff
@@ -539,7 +664,19 @@ ProgramDescriptor build_cross_core_program_descriptor(
...
+    // UINT16 keys must also run in fp32_dest_acc_en mode so the SFPU compares
+    // them as 32-bit floats.  See comment in SortProgramFactorySingleRowSingleCore
+    // for the full rationale.
+    const bool is_32_bit_data = is_32_bit_index || input_tensor_cb_data_format == tt::DataFormat::Float32 ||
+                                input_tensor_cb_data_format == tt::DataFormat::UInt16;
+
+    // UINT16 inputs are rejected in validate_on_program_cache_miss (Wt <= SORT_WT_THRESHOLD),
+    // so they are always routed to SortProgramFactorySingleRowSingleCore before reaching
+    // this factory.  The assertion below guards against that invariant being broken.
+    TT_ASSERT(
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/sort/device/sort_program_factory.cpp]: Two things: the rationale above is inaccurate — `validate_on_program_cache_miss` does not reject UINT16; `select_program_factory` is what keeps it out of here (`!is_uint16 && Wt <= total_number_of_tiles_for_hybrid_approach`). And `TT_ASSERT` is compiled out in release builds, so if that routing invariant is ever broken, release builds sort UINT16 keys through a factory with no conversion path and silently return wrong values instead of failing. `TT_FATAL` is the right guard for this one.

File: ttnn/cpp/ttnn/operations/data_movement/sort/device/sort_program_factory.cpp [hunk 3]
```diff
@@ -222,6 +258,74 @@ ProgramDescriptor SortProgramFactorySingleRowSingleCore::create_descriptor(
         });
     }
 
+    // UInt16 conversion CB: writer reads Float32 sorted values from c_4 and uses this
+    // single-tile UInt16 CB as an intermediate buffer to convert Float32 → UInt16 before
+    // DMA'ing to DRAM.  Allocated for all inputs so the writer can reference c_11 via
+    // get_tile_size() at compile time regardless of input dtype.
+    constexpr uint32_t uint16_conv_cb_index = tt::CBIndex::c_11;
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/sort/device/sort_program_factory.cpp]: c_11-c_14 are allocated for every input dtype, so every bf16/fp32 sort now pays 2 UInt16 tiles + 2 * W_value_bytes of L1 it never touches (~20 KB for fp32 RM at Wt=64). Given this factory is already L1-tight enough to need `SORT_WT_THRESHOLD_UINT16_ROW_MAJOR`, worth gating all four on `is_uint16_input`. The `get_tile_size()` argument does not force unconditional allocation: `reader_single_row_multi_core.cpp` declares `uint16_stage_tile_size` inside its `if constexpr (is_uint16_fp32_mode)` block — doing the same in the single-core reader/writer drops the unconditional reference.

**Review by Aswinmcw — APPROVED** [review]
Approving. Please fill in the Summary and link #46331, and state in the description that UINT16 inputs now return UINT32 indices (a UINT16 `indices` output is now a hard error) — that is a user-visible output-spec change.

**Review by mgajewskiTT — APPROVED** [review]

**Review by VirdhatchaniKN — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]
